### PR TITLE
Have #0 operator preserve a range's given bounds

### LINKED ISSUE #5165: Count operator always returns 1..0 when count is zero
## Bug Report
The range count operator ignores the range's low bound when count is 0, and always returns 1..0.  It would be better if the low bound would be kept so that the count operator can produce empty ranges with a specified low bound.

### Steps to Reproduce

writeln(0..#0); // gets: 1..0, want: 0..-1

### Configuration Information
chpl Version 1.14.0

### LINKED ISSUE #13484: Domain created with dimensions {0..#0} becomes {1..0}
It seems that creating an empty range will force it to become `1..0`. 

```chpl
writeln(0..#0);
```
[TIO](https://tio.run/##S85ILEjN@f@/vCizJDUnT8NAT0/ZQNP6/38A)

```
1..0
```

This applies to domains, which makes it very tricky when assuming that an empty domain is 0-based when they set it as such.

### PR DESCRIPTION
**Pull request by bradcray**
Prior to this change, applying #0 to a range would always result
in the range 1..0 coming back.  This meant that, for example,
`42..#0` would return `1..0` where arguably the answer `42..41`
would be better since it preserves the given bound (and similarly,
`10..20#0` should give `10..9`.  Here, I reimplement the #0
case to preserve the appropriate bound of the original range.

One complicated case is the one where computing the "next element"
in the direction that results in an empty range wraps around the
legal range of idxType.  For example, (0:uint)..#0 would like
to return 0..-1 but since -1 can't be represented, it's not clear
what should happen.  In discussion on issue #5165, we decided to
apply normal integral math and have the normal behavior occur
(which suggests uint values will wrap around and int values can't
be depended upon).

Here, I make that change and also update existing tests to reflect it.
Test `test/types/range/hilde/count.chpl` is quite possibly non-portable
because of its use of maximal / minimal signed integers, but I thought
I'd check it in and see how non-portable it is for the compilers we test
with before retiring it prematurely (I did add a comment to the source
to point it out).

Resolves #13484.
Resolves #5165 (with relevant conversation about maximal vs. empty ranges split off to issue #13717).

### CODE REVIEW

File: modules/internal/ChapelRange.chpl [hunk 1]
```diff
@@ -1555,16 +1555,37 @@ proc _cast(type t, r: range(?)) where isRangeType(t) {
...
-                       _stride = r.stride,
-                       _alignment = r.chpl_intToIdx(0),
-                       _aligned = false);
+    proc absSameType(str) {
+      if (r.stride < 0) {
+        return (-r.stride):resultType;
+      } else {
+        return r.stride:resultType;
+      }
+    }
```
Comment (vasslitvinov) [inline modules/internal/ChapelRange.chpl]: absSameType() is not using its formal. Please either remove the formal or switch to using it.

File: test/types/range/hilde/count.good [hunk 2]
```diff
@@ -3,42 +3,42 @@
...
-1..0 by -1073741824
-1..0 by 1073741825
+-2147483648..2147483647
+-2147483648..2147483646 by 2
+-2147483648..1073741824 by -1073741824 align 1073741823
+-2147483648..1073741823 by 1073741825
+-9223372036854775808..9223372036854775807
+-9223372036854775808..9223372036854775806 by 2
+-9223372036854775808..9223372035781033984 by -1073741824 align 1073741823
+-9223372036854775808..9223372035781033983 by 1073741825
```
Comment (vasslitvinov) [inline test/types/range/hilde/count.good]: I am suspicious about these added `align` clauses. However, if you looked at them and found them correct, that's fine by me.

File: modules/internal/ChapelRange.chpl (hunk 1 above)
Comment (bradcray) [reply modules/internal/ChapelRange.chpl]: D'oh!  Thanks for catching that!

File: test/types/range/hilde/count.good (hunk 2 above)
Comment (bradcray) [reply test/types/range/hilde/count.good]: I've double-checked them and believe them to be correct.  That said, this last block of changes are likely to be non-portable due to the reliance on integer wraparound, so arguably should be removed.  I thought I'd check them in to see how non-portable they were before retiring them prematurely, though.

Comment (vasslitvinov) [reply test/types/range/hilde/count.good]: Sounds good.

**Pull request merged**